tests/db_tests/test_queries.py

```diff
@@ -481,3 +484,27 @@ def test_project_if_no_outcomes(seeded_test_client, additional_test_data):
 
     test_df = pd.read_sql(test_query_proj.statement, con=db.engine.connect())
     assert list(test_df["project_name"])
+
+
+def test_with_transaction_max_retries(mocker, test_client):
+    """
+    Test the behavior of 'with_transaction' function in case of IntegrityError.
+
+    This test simulates a scenario where a function raises an IntegrityError
+    and ensures that the 'with_transaction' function retries the operation up to the
+    specified maximum number of times. It also checks whether the error is properly
+    logged after reaching the maximum number of retries.
+    """
+    max_retries = 3
+    sleep_duration = 0.1
+    error_type = exc.IntegrityError
+
+    mocked_function = mocker.Mock(side_effect=exc.IntegrityError("Mocked IntegrityError", None, None))
+    mocked_function.__name__ = "mocked_function"
+    with patch.object(current_app.logger, "error") as mock_logger:
```

Comment:
caplog looks like a cleaner way of asserting on log contents

see this example:
https://github.com/communitiesuk/funding-service-design-post-award-data-store/blob/9210b8ad2400322305a45c9eafc20a89259d70c2/tests/test_handlers.py#L10-L24